db/version_set.cc

```diff
@@ -1674,13 +1673,132 @@ void VersionStorageInfo::GetOverlappingInputs(
   }
 }
 
+// Store in "*inputs" files in "level" that within range [begin,end]
+// Guarantee a "clean cut" boundary between the files in inputs
+// and the surrounding files and the maxinum number of files.
+// This will ensure that no parts of a key are lost during compaction.
+// If hint_index is specified, then it points to a file in the range.
+// The file_index returns a pointer to any file in an overlapping range.
+void VersionStorageInfo::GetCleanInputsWithinInterval(
+    int level, const InternalKey* begin, const InternalKey* end,
+    std::vector<FileMetaData*>* inputs, int hint_index, int* file_index) const {
+  if (level >= num_non_empty_levels_) {
+    // this level is empty, no inputs within range
+    return;
+  }
+
+  inputs->clear();
+  Slice user_begin, user_end;
+  if (begin != nullptr) {
+    user_begin = begin->user_key();
+  }
+  if (end != nullptr) {
+    user_end = end->user_key();
+  }
+  if (file_index) {
+    *file_index = -1;
+  }
+  const Comparator* user_cmp = user_comparator_;
+  if (begin != nullptr && end != nullptr && level > 0) {
+    GetOverlappingInputsRangeBinarySearch(level, user_begin, user_end, inputs,
+                                     hint_index, file_index, true /* within_interval */);
+    return;
+  }
+
+  const FdWithKeyRange* files = level_files_brief_[level].files;
+  size_t num_files = level_files_brief_[level].num_files;
+  std::vector<size_t> indices(num_files);
+  for (size_t i = 0; i < num_files; i++) {
+    indices[i] = i;
+  }
+  sort(indices.begin(), indices.end(), [&files, user_cmp](size_t i, size_t j) {
+      return user_cmp->Compare(ExtractUserKey(files[i].smallest_key), ExtractUserKey(files[j].smallest_key)) < 0;
+    });
+  std::vector<std::pair<Slice, Slice>> overlapping_range_list;
+  std::vector<std::vector<size_t>> files_in_range_list;
+  for (size_t i = 0; i < num_files; i++) {
+    const FdWithKeyRange* f = &(files[indices[i]]);
+    const Slice file_start = ExtractUserKey(f->smallest_key);
+    const Slice file_limit = ExtractUserKey(f->largest_key);
+    if (begin != nullptr && user_cmp->Compare(file_limit, user_begin) < 0) {
+      // "f" is completely before specified range; skip it
+    } else if (end != nullptr && user_cmp->Compare(file_start, user_end) > 0) {
+      // "f" is completely after specified range; break the loop
+      break;
+    } else {
+      if (overlapping_range_list.empty() ||
+          user_cmp->Compare(file_start, overlapping_range_list.back().second) >
+              0) {
+        overlapping_range_list.push_back(std::pair<Slice,Slice>(file_start, file_limit));
```

Comment:
I think I mentioned this offline. Handling L0 is not needed in this case. This function will be significantly simplified if files won't overlap.